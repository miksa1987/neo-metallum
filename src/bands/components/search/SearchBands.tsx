import React, { useState } from 'react';
import { useQuery } from '@apollo/react-hooks';
import { gql } from 'apollo-boost';

import { Loader } from '../../../common';
import { BandType } from '../../../common/types';
import { BandsList } from '../common/BandsList';
import { Fields } from './Fields';
import { FieldType } from './Field';

export const GET_BANDS_QUERY = gql`
  query findBandsByLetter(
    $name: String!, 
    $start: Int!, 
    $genre: String!,
    $formationFrom: String!, 
    $formationTo: String!
  ) {
    bands(
      name: $name, 
      start: $start,
      genre: $genre,
      formationFrom: $formationFrom,
      formationTo: $formationTo
    ) {
      id
      name
      country
      genre
    }
  }
`;

export const SearchBands = () => {
  const [ variables, setVariables ] = useState({
    name: '',
    genre: '',
    formationFrom: '',
    formationTo: '',
    start: 0
  });

  const { data, loading, error } = useQuery(GET_BANDS_QUERY, { variables });

  const fields: FieldType[] = [
    {
      name: 'name',
      component: 'input',
      type: 'text',
      placeholder: 'Band name'
    },
    {
      name: 'genre',
      component: 'input',
      type: 'text',
      placeholder: 'Band genre'
    },
    {
      name: 'formationFrom',
      component: 'input',
      type: 'text',
      placeholder: 'Band formation from'
    },
    {
      name: 'formationTo',
      component: 'input',
      type: 'text',
      placeholder: 'Band formation to'
    }
  ];

  if (loading) {
    return (
      <div>
        <Fields fields={fields} setValues={setVariables} />
        <Loader />
      </div>
    );
  }

  const bands: BandType[] = data.bands;

  return (
    <div>
      <Fields fields={fields} setValues={setVariables} />
      <BandsList bands={bands} />
    </div>
  );
}