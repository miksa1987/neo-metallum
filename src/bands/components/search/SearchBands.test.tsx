import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, cleanup, fireEvent, act } from '@testing-library/react';
import {Â BrowserRouter } from 'react-router-dom';
import { MockedProvider } from '@apollo/react-testing';
import { gql } from 'apollo-boost'; // For the query to work

import { SearchBands, GET_BANDS_QUERY } from './SearchBands';

const mockUseQuery = jest.fn(() => ({
  data: {
    bands: [
      {
        name: 'testii',
        country: 'finland',
        id: '1',
        genre: 'thrash metal'
      }
    ]
  }
}));

let queryparams: any = null;
jest.mock('@apollo/react-hooks', () => ({
  ...jest.requireActual('@apollo/react-hooks'),
  useQuery: (query: any, params: any) => {
    queryparams = params;

    return {
      data: {
        bands: [
          {
            name: 'testii',
            country: 'finland',
            id: '1',
            genre: 'thrash metal'
          }
        ]
      }
    };
  }
}));

describe('SearchBands', () => {
  afterEach(cleanup);

  test('Renders correctly', () => {
    const component = render(
      <BrowserRouter>
        <SearchBands />
      </BrowserRouter>
    );
    component.debug();

    expect(component).toMatchSnapshot();
  });

  // I don't know wtf I'm doing 
  test('Query is called with correct variables', () => {
    mockUseQuery.mockClear();

    const component = render(
      <BrowserRouter>
        <SearchBands />
      </BrowserRouter>
    );
      
    const nameInput: any = component.container.querySelector('#name');
    const genreInput: any = component.container.querySelector('#genre');
    const searchButton: any = component.container.querySelector('#submit-search');

    fireEvent.change(nameInput, { target: { value: 'testii' }});
    fireEvent.change(genreInput, { target: { value: 'thrash' }});
    fireEvent.click(searchButton);

    const expectedResult = {
      variables: {
        name: 'testii',
        genre: 'thrash',
        formationFrom: '',
        formationTo: '',
        start: 0
      }
    };

    expect(queryparams).toEqual(expectedResult);
  });
});