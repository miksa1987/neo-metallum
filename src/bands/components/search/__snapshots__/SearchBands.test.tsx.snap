// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBands Renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="sc-fznZeY crAIIR"
        >
          <form>
            <div
              class="sc-fznKkj iRpFma"
            >
              <strong>
                Band name
              </strong>
              <input
                id="name"
                name="name"
                placeholder="Band name"
                type="text"
                value=""
              />
            </div>
            <div
              class="sc-fznKkj iRpFma"
            >
              <strong>
                Band genre
              </strong>
              <input
                id="genre"
                name="genre"
                placeholder="Band genre"
                type="text"
                value=""
              />
            </div>
            <div
              class="sc-fznKkj iRpFma"
            >
              <strong>
                Band formation from
              </strong>
              <input
                id="formationfrom"
                name="formationFrom"
                placeholder="Band formation from"
                type="text"
                value=""
              />
            </div>
            <div
              class="sc-fznKkj iRpFma"
            >
              <strong>
                Band formation to
              </strong>
              <input
                id="formationto"
                name="formationTo"
                placeholder="Band formation to"
                type="text"
                value=""
              />
            </div>
            <button
              id="submit-search"
              type="submit"
            >
              Search
            </button>
          </form>
        </div>
        <div
          class="sc-AxirZ enylqG"
        >
          <div
            class="sc-fznyAO ePjXxu"
          >
            <div
              class=""
              style="width: 33.333333333333336%;"
            >
              <div>
                <div
                  class="sc-fzplWN hDAGEZ"
                >
                  <a
                    href="/band/1"
                    id="band-info-link"
                  >
                    <h2>
                      testii
                    </h2>
                  </a>
                  <img
                    alt="band"
                    class="sc-AxiKw hPbuBq"
                    src="/images/band/10.jpg"
                  />
                  <strong>
                    finland
                  </strong>
                  <p>
                    thrash metal
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="sc-fznZeY crAIIR"
      >
        <form>
          <div
            class="sc-fznKkj iRpFma"
          >
            <strong>
              Band name
            </strong>
            <input
              id="name"
              name="name"
              placeholder="Band name"
              type="text"
              value=""
            />
          </div>
          <div
            class="sc-fznKkj iRpFma"
          >
            <strong>
              Band genre
            </strong>
            <input
              id="genre"
              name="genre"
              placeholder="Band genre"
              type="text"
              value=""
            />
          </div>
          <div
            class="sc-fznKkj iRpFma"
          >
            <strong>
              Band formation from
            </strong>
            <input
              id="formationfrom"
              name="formationFrom"
              placeholder="Band formation from"
              type="text"
              value=""
            />
          </div>
          <div
            class="sc-fznKkj iRpFma"
          >
            <strong>
              Band formation to
            </strong>
            <input
              id="formationto"
              name="formationTo"
              placeholder="Band formation to"
              type="text"
              value=""
            />
          </div>
          <button
            id="submit-search"
            type="submit"
          >
            Search
          </button>
        </form>
      </div>
      <div
        class="sc-AxirZ enylqG"
      >
        <div
          class="sc-fznyAO ePjXxu"
        >
          <div
            class=""
            style="width: 33.333333333333336%;"
          >
            <div>
              <div
                class="sc-fzplWN hDAGEZ"
              >
                <a
                  href="/band/1"
                  id="band-info-link"
                >
                  <h2>
                    testii
                  </h2>
                </a>
                <img
                  alt="band"
                  class="sc-AxiKw hPbuBq"
                  src="/images/band/10.jpg"
                />
                <strong>
                  finland
                </strong>
                <p>
                  thrash metal
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
